File: packages/entity-extraction/ios/Plugin/Classes/Results/ExtractEntitiesResult.swift
import Foundation
import Capacitor
import MLKitEntityExtraction

@objc public class ExtractEntitiesResult: NSObject, Result {
    let text: String
    let annotations: [EntityAnnotation]

    init(text: String, annotations: [EntityAnnotation]) {
        self.text = text
        self.annotations = annotations
    }

    @objc public func toJSObject() -> AnyObject {
        let nsText = text as NSString
        var annotationsResult = JSArray()
        for annotation in annotations {
            annotationsResult.append(ExtractEntitiesResult.createAnnotationResult(annotation, text: nsText))
        }
        var result = JSObject()
        result["annotations"] = annotationsResult
        return result as AnyObject
    }

    private static func createAnnotationResult(_ annotation: EntityAnnotation, text: NSString) -> JSObject {
        var entitiesResult = JSArray()
        for entity in annotation.entities {
            entitiesResult.append(ExtractEntitiesResult.createEntityResult(entity))
        }
        var result = JSObject()
        result["text"] = text.substring(with: annotation.range)
        result["start"] = annotation.range.location
        result["end"] = annotation.range.location + annotation.range.length
        result["entities"] = entitiesResult
        return result
    }

    private static func createEntityResult(_ entity: Entity) -> JSObject {
        var result = JSObject()
        result["type"] = EntityExtractionHelper.convertEntityTypeToString(entity.entityType)

        if let dateTimeEntity = entity.dateTimeEntity {
            result["dateTimeGranularity"] = EntityExtractionHelper.convertDateTimeGranularityToString(
                dateTimeEntity.dateTimeGranularity)
            result["timestamp"] = Int(dateTimeEntity.dateTime.timeIntervalSince1970 * 1000)
        }

        if let moneyEntity = entity.moneyEntity {
            result["unnormalizedCurrency"] = moneyEntity.unnormalizedCurrency
            result["integerPart"] = moneyEntity.integerPart
            result["fractionalPart"] = moneyEntity.fractionalPart
        }

        if let paymentCardEntity = entity.paymentCardEntity {
            result["paymentCardNetwork"] = EntityExtractionHelper.convertPaymentCardNetworkToString(
                paymentCardEntity.paymentCardNetwork)
            result["paymentCardNumber"] = paymentCardEntity.paymentCardNumber
        }

        if let flightNumberEntity = entity.flightNumberEntity {
            result["airlineCode"] = flightNumberEntity.airlineCode
            result["flightNumber"] = flightNumberEntity.flightNumber
        }

        if let isbnEntity = entity.isbnEntity {
            result["isbn"] = isbnEntity.isbn
        }

        if let ibanEntity = entity.ibanEntity {
            result["iban"] = ibanEntity.iban
            result["ibanCountryCode"] = ibanEntity.countryCode
        }

        if let trackingNumberEntity = entity.trackingNumberEntity {
            result["parcelCarrier"] = EntityExtractionHelper.convertParcelCarrierToString(
                trackingNumberEntity.parcelCarrier)
            result["parcelTrackingNumber"] = trackingNumberEntity.parcelTrackingNumber
        }

        return result
    }
}

File: packages/entity-extraction/android/src/main/java/io/capawesome/capacitorjs/plugins/mlkit/entityextraction/classes/results/ExtractEntitiesResult.java
package io.capawesome.capacitorjs.plugins.mlkit.entityextraction.classes.results;

import androidx.annotation.NonNull;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.google.mlkit.nl.entityextraction.DateTimeEntity;
import com.google.mlkit.nl.entityextraction.Entity;
import com.google.mlkit.nl.entityextraction.EntityAnnotation;
import com.google.mlkit.nl.entityextraction.FlightNumberEntity;
import com.google.mlkit.nl.entityextraction.IbanEntity;
import com.google.mlkit.nl.entityextraction.IsbnEntity;
import com.google.mlkit.nl.entityextraction.MoneyEntity;
import com.google.mlkit.nl.entityextraction.PaymentCardEntity;
import com.google.mlkit.nl.entityextraction.TrackingNumberEntity;
import io.capawesome.capacitorjs.plugins.mlkit.entityextraction.EntityExtractionHelper;
import io.capawesome.capacitorjs.plugins.mlkit.entityextraction.interfaces.Result;
import java.util.List;

public class ExtractEntitiesResult implements Result {

    @NonNull
    private final List<EntityAnnotation> annotations;

    public ExtractEntitiesResult(@NonNull List<EntityAnnotation> annotations) {
        this.annotations = annotations;
    }

    @Override
    @NonNull
    public JSObject toJSObject() {
        JSArray annotationsArray = new JSArray();
        for (EntityAnnotation annotation : annotations) {
            annotationsArray.put(createAnnotationResult(annotation));
        }
        JSObject result = new JSObject();
        result.put("annotations", annotationsArray);
        return result;
    }

    @NonNull
    private static JSObject createAnnotationResult(@NonNull EntityAnnotation annotation) {
        JSArray entitiesArray = new JSArray();
        for (Entity entity : annotation.getEntities()) {
            entitiesArray.put(createEntityResult(entity));
        }
        JSObject result = new JSObject();
        result.put("text", annotation.getAnnotatedText());
        result.put("start", annotation.getStart());
        result.put("end", annotation.getEnd());
        result.put("entities", entitiesArray);
        return result;
    }

    @NonNull
    private static JSObject createEntityResult(@NonNull Entity entity) {
        JSObject result = new JSObject();
        result.put("type", EntityExtractionHelper.convertEntityTypeToString(entity.getType()));

        DateTimeEntity dateTimeEntity = entity.asDateTimeEntity();
        if (dateTimeEntity != null) {
            result.put(
                "dateTimeGranularity",
                EntityExtractionHelper.convertDateTimeGranularityToString(dateTimeEntity.getDateTimeGranularity())
            );
            result.put("timestamp", dateTimeEntity.getTimestampMillis());
        }

        MoneyEntity moneyEntity = entity.asMoneyEntity();
        if (moneyEntity != null) {
            result.put("unnormalizedCurrency", moneyEntity.getUnnormalizedCurrency());
            result.put("integerPart", moneyEntity.getIntegerPart());
            result.put("fractionalPart", moneyEntity.getFractionalPart());
        }

        PaymentCardEntity paymentCardEntity = entity.asPaymentCardEntity();
        if (paymentCardEntity != null) {
            result.put(
                "paymentCardNetwork",
                EntityExtractionHelper.convertPaymentCardNetworkToString(paymentCardEntity.getPaymentCardNetwork())
            );
            result.put("paymentCardNumber", paymentCardEntity.getPaymentCardNumber());
        }

        FlightNumberEntity flightNumberEntity = entity.asFlightNumberEntity();
        if (flightNumberEntity != null) {
            result.put("airlineCode", flightNumberEntity.getAirlineCode());
            result.put("flightNumber", flightNumberEntity.getFlightNumber());
        }

        IsbnEntity isbnEntity = entity.asIsbnEntity();
        if (isbnEntity != null) {
            result.put("isbn", isbnEntity.getIsbn());
        }

        IbanEntity ibanEntity = entity.asIbanEntity();
        if (ibanEntity != null) {
            result.put("iban", ibanEntity.getIban());
            result.put("ibanCountryCode", ibanEntity.getIbanCountryCode());
        }

        TrackingNumberEntity trackingNumberEntity = entity.asTrackingNumberEntity();
        if (trackingNumberEntity != null) {
            result.put("parcelCarrier", EntityExtractionHelper.convertParcelCarrierToString(trackingNumberEntity.getParcelCarrier()));
            result.put("parcelTrackingNumber", trackingNumberEntity.getParcelTrackingNumber());
        }

        return result;
    }
}
